src/platform/PICO/uart/uart_pio.c

```diff
@@ -0,0 +1,374 @@
+        }
+
+        return txProgram_offset >= 0;
+   } else {
+        if (rxProgram_offset < 0) {
+            rxProgram_offset = pio_add_program(pio, program);
+        }
+
+        return rxProgram_offset >= 0;
+    }
+}
+
+#if SERIAL_PIOUART_COUNT > 0
+static void uartPioIrqHandler(uartPort_t *s, pioDetails_t *pioDetailsPtr)
+{
+    io_rw_32 *enableRegPtr = pioDetailsPtr->enableReg;
+    io_ro_32 *statusRegPtr = pioDetailsPtr->statusReg;
+    // Is RX enabled and does irq status indicate RX fifo not empty?
+    if (*enableRegPtr & pioDetailsPtr->rx_intr_bit && *statusRegPtr & pioDetailsPtr->rx_intr_bit) {
+        uint sm_rx = pioDetailsPtr->sm_rx;
+
+        // 8-bit read from the uppermost byte of the FIFO, as data is left-justified
+        io_ro_32 *rxfifo_shift = &uartPio->rxf[sm_rx];
+        serialReceiveCallbackPtr rxCallback = s->port.rxCallback;
+        if (rxCallback) {
+            void *rxCallbackData = s->port.rxCallbackData;
+            while (!pio_sm_is_rx_fifo_empty(uartPio, sm_rx)) {
+                const uint8_t ch = *rxfifo_shift >> 24; // left aligned, 8 bits
+                rxCallback(ch, rxCallbackData);
+            }
+        } else {
+            volatile uint8_t *rxBuffer = s->port.rxBuffer;
+            uint32_t rxBufferSize = s->port.rxBufferSize;
+            while (!pio_sm_is_rx_fifo_empty(uartPio, sm_rx)) {
+                const uint8_t ch = *rxfifo_shift >> 24; // left aligned, 8 bits
+                rxBuffer[s->port.rxBufferHead] = ch;
+                s->port.rxBufferHead = (s->port.rxBufferHead + 1) % rxBufferSize;
+            }
+        }
+    }
+
+    // Is TX enabled?
+    if (*enableRegPtr & pioDetailsPtr->tx_intr_bit) {
+        uint sm_tx = pioDetailsPtr->sm_tx;
+        while (!pio_sm_is_tx_fifo_full(uartPio, sm_tx)) {
+            if (s->port.txBufferTail != s->port.txBufferHead) {
+                ///bprintf("uartIrqHandler PIO TX put %x", s->port.txBuffer[s->port.txBufferTail]);
+                // Send 8 bits of data in a 32-bit word. The PIO program will emit the low 8 bits.
+                pio_sm_put(uartPio, sm_tx, s->port.txBuffer[s->port.txBufferTail]);
+                s->port.txBufferTail = (s->port.txBufferTail + 1) % s->port.txBufferSize;
+            } else {
+                // Done sending the buffer, clear the pio TX interrupt enable
+                //bprintf("uart pio done put %d, disabling tx interrupt",c);
+                int irqn_index = PIO_IRQ_INDEX(pioDetailsPtr->irqn);
+                pio_interrupt_source_t irqSourceTX = pio_get_tx_fifo_not_full_interrupt_source(sm_tx);
+                pio_set_irqn_source_enabled(uartPio, irqn_index, irqSourceTX, false);
+                break;
+            }
+        }
+    }
+}
+#endif
+
+static void on_pioUART0(void)
+{
+///    bprintf("\n\n on_pioUART0");
+#ifdef USE_PIOUART0
+    uartPioIrqHandler(&pioUartDevice[0].port, UART_PIO_DETAILS_PTR(SERIAL_PORT_PIOUART0));
```

Comment:
Not sure here, but indirection is not necessary:
```suggestion
    uartPioIrqHandler(&pioUartDevice[0].port, &uartPioDetails[0]);
```